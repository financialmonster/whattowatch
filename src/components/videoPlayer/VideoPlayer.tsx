import React, { FC,  memo, forwardRef } from 'react';

import { getTimeFromMins, isRefInitialized } from 'utils';
import { useVideoPlayerProgress } from 'hooks/useVideoPlayerProgress';
import { useActiveVideoPlayer } from 'hooks/useActiveVideoPlayer';

type TVideoPlayerProps = {
    src: string | undefined,
    posterImage: string | undefined,
    runTime: number | undefined,
    ref: React.RefObject<HTMLVideoElement>,
    isPlaying: boolean,
    handleExitBtnClick: () => void
}

export const VideoPlayer: FC<TVideoPlayerProps> = memo(forwardRef((props, ref) => {
    const {handleExitBtnClick, src, posterImage, runTime, isPlaying} = props;

    const progress = useVideoPlayerProgress(ref);
    const {isActive, playPauseBtnHandler, exitBtnClickHandler} =
        useActiveVideoPlayer(isPlaying, ref, handleExitBtnClick);

    const playerStyle = (isPlaying)
        ? {}
        : {display: `none`};

    const fullScreenBtnClickHandler = () => {
        if(isRefInitialized(ref)) {
            ref.current.requestFullscreen();
        }
    }

    return (
        <div className="player" style={playerStyle}>
            <video className="player__video" src={src} poster={posterImage} ref={ref}></video>
            <button type="button" className="player__exit" onClick={exitBtnClickHandler}>
                Exit
            </button>
            <div className="player__controls">
                <div className="player__controls-row">
                    <div className="player__time">
                        <progress className="player__progress" value={progress} max="100"></progress>
                        <div className="player__toggler" style={{ left: `${progress}%` }}>
                            Toggler
                        </div>
                    </div>
                    <div className="player__time-value">{getTimeFromMins(runTime)}</div>
                </div>
                <div className="player__controls-row">
                    <button type="button" className="player__play" onClick={playPauseBtnHandler}>
                        <svg viewBox={(isActive) ? `0 0 14 21` : `0 0 320.001 320.001`} width="14" height="21">
                            <use xlinkHref={(isActive) ? `#pause` : "#play"}></use>
                        </svg>
                        <span>{(isActive) ? `Pause` : `Play`}</span>
                    </button>
                    <div className="player__name">Transpotting</div>
                    <button type="button" className="player__full-screen" onClick={fullScreenBtnClickHandler}>
                        <svg viewBox="0 0 27 27" width="27" height="27">
                            <use xlinkHref="#full-screen"></use>
                        </svg>
                        <span>Full screen</span>
                    </button>
                </div>
            </div>
        </div>
    );
}));
