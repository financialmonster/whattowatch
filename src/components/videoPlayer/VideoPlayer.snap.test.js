import React from 'react';
import renderer from 'react-test-renderer';

import { VideoPlayer } from './VideoPlayer';

it(`VideoPlayer: correctly rendered`, () => {
    const tree = renderer.create(
        <VideoPlayer />    
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
