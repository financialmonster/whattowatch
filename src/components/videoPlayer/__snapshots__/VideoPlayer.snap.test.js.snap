// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer: correctly rendered 1`] = `
<div
  className="player"
  style={
    Object {
      "display": "none",
    }
  }
>
  <video
    className="player__video"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    title="Exit"
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        12:00:00
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        title="Play"
        type="button"
      >
        <svg
          height="21"
          viewBox="0 0 320.001 320.001"
          width="14"
        >
          <use
            xlinkHref="#play"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Transpotting
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        title="Full screen"
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
